File: bot.py
import discord
from discord.ext import tasks
import requests
import json
from datetime import datetime
import aiohttp
import asyncio
from bs4 import BeautifulSoup
from selenium import webdriver
from selenium.webdriver.chrome.service import Service as ChromeService
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from webdriver_manager.chrome import ChromeDriverManager
import re
import os
from config import UPDATE_INTERVAL, WEBHOOK_URL, SERVER_ID

class ServerStatsBot:
    def __init__(self, webhook_url, server_id):
        self.webhook_url = webhook_url
        self.server_id = server_id
        self.stats_file = 'stats_db.json'
        self.stats_db = self.load_stats()
        self.session = None
        self.driver = None
        self.hourly_update_task = None
        self.update_interval = UPDATE_INTERVAL
        self.setup_driver()

    async def start(self):
        self.session = aiohttp.ClientSession()
        try:
            print("Führe erste Aktualisierung durch...")
            await self.hourly_update()
            
            print(f"Starte Update-Task (Intervall: {self.update_interval} Sekunden)...")
            if not self.hourly_update_task:
                # Erstelle den Task mit dem korrekten Intervall
                self.hourly_update.change_interval(seconds=self.update_interval)
                self.hourly_update_task = self.hourly_update.start()
            
            print("Bot läuft - warte auf Updates...")
            print("Drücke STRG+C zum Beenden")
            
            while True:
                await asyncio.sleep(60)
                
        except asyncio.CancelledError:
            print("Bot-Task wurde abgebrochen")
        except Exception as e:
            print(f"Fehler im Bot: {e}")
        finally:
            print("Räume Bot-Ressourcen auf...")
            if self.session and not self.session.closed:
                await self.session.close()
            if self.driver:
                self.driver.quit()

    def setup_driver(self):
        try:
            if self.driver:
                self.driver.quit()
                
            chrome_options = Options()
            chrome_options.add_argument('--headless=new')
            chrome_options.add_argument('--no-sandbox')
            chrome_options.add_argument('--disable-dev-shm-usage')
            chrome_options.add_argument('--disable-gpu')
            chrome_options.add_argument('--window-size=1920,1080')
            chrome_options.add_argument('--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36')
            
            service = ChromeService(ChromeDriverManager().install())
            self.driver = webdriver.Chrome(
                options=chrome_options,
                service=service
            )
            self.driver.implicitly_wait(10)
            print("Chrome-Driver erfolgreich initialisiert")
            
        except Exception as e:
            print(f"Fehler beim Setup des Drivers: {e}")
            if self.driver:
                self.driver.quit()
                self.driver = None

    def get_player_count(self):
        try:
            if not self.driver:
                print("Driver nicht verfügbar, initialisiere neu...")
                self.setup_driver()
                
            if not self.driver:
                print("Driver konnte nicht initialisiert werden")
                return 0
                
            # Seite neu laden und warten bis sie vollständig geladen ist
            self.driver.get('https://servers.redm.gg/servers/detail/bzy79l')
            WebDriverWait(self.driver, 20).until(
                EC.presence_of_element_located((By.TAG_NAME, "body"))
            )
            
            # Warte auf das spezifische Element mit der Spielerzahl
            player_element = WebDriverWait(self.driver, 20).until(
                EC.presence_of_element_located((By.CSS_SELECTOR, 'div.connect-bar div.right span.material-icons-outlined'))
            )
            
            print("Gefundenes Element:", player_element.text)
            
            # Finde das benachbarte Element mit der tatsächlichen Zahl
            player_count = self.driver.find_element(By.CSS_SELECTOR, 'div.connect-bar div.right').text
            print("Vollständiger Text:", player_count)
            
            # Extrahiere die erste Zahl aus dem Text
            numbers = re.findall(r'\d+', player_count)
            if numbers:
                return int(numbers[0])
            return 0
                
        except Exception as e:
            print(f"Fehler beim Scraping: {e}")
            # Bei Fehlern Driver neu initialisieren
            self.setup_driver()
            return 0

    def load_stats(self):
        if os.path.exists(self.stats_file):
            try:
                with open(self.stats_file, 'r') as f:
                    data = json.load(f)
                    # Konvertiere Timestamp-Strings zurück zu datetime-Objekten
                    for entry in data['hourly']:
                        entry['timestamp'] = datetime.fromisoformat(entry['timestamp'])
                    return data
            except Exception as e:
                print(f"Fehler beim Laden der Statistiken: {e}")
        return {'hourly': []}

    def save_stats(self):
        try:
            # Konvertiere datetime-Objekte zu Strings für JSON
            data = {'hourly': []}
            for entry in self.stats_db['hourly']:
                data['hourly'].append({
                    'timestamp': entry['timestamp'].isoformat(),
                    'players': entry['players']
                })
            with open(self.stats_file, 'w') as f:
                json.dump(data, f)
        except Exception as e:
            print(f"Fehler beim Speichern der Statistiken: {e}")

    @tasks.loop(seconds=UPDATE_INTERVAL)
    async def hourly_update(self):
        try:
            print(f"\nFühre Update aus - {datetime.now().strftime('%H:%M:%S')}")
            players = self.get_player_count()
            timestamp = datetime.now()
            
            # Neue Daten speichern
            self.stats_db['hourly'].append({
                'timestamp': timestamp,
                'players': players
            })
            
            # Statistiken berechnen
            stats = self.calculate_stats()
            
            embed = discord.Embed(
                title="Misty Mountain - Aktuelle Spielerzahl",
                description=f"**{players}** Spieler online",
                color=discord.Color.blue(),
                timestamp=timestamp
            )
            
            # Statistiken hinzufügen
            embed.add_field(
                name="24 Stunden",
                value=f"Durchschnitt: {stats['day_avg']}\nMaximum: {stats['day_max']}\nMinimum: {stats['day_min']}",
                inline=True
            )
            embed.add_field(
                name="7 Tage",
                value=f"Durchschnitt: {stats['week_avg']}\nMaximum: {stats['week_max']}\nMinimum: {stats['week_min']}",
                inline=True
            )
            embed.add_field(
                name="30 Tage",
                value=f"Durchschnitt: {stats['month_avg']}\nMaximum: {stats['month_max']}\nMinimum: {stats['month_min']}",
                inline=True
            )
            
            await self.send_webhook(embed)
            
            # Speichere die Statistiken nach dem Update
            self.save_stats()
            
            # Entferne alte Einträge (älter als 30 Tage)
            now = datetime.now()
            self.stats_db['hourly'] = [
                entry for entry in self.stats_db['hourly']
                if (now - entry['timestamp']).total_seconds() <= 2592000
            ]
            
            print(f"Update abgeschlossen - {players} Spieler online")
            
        except Exception as e:
            print(f"Fehler beim Update: {e}")

    def calculate_stats(self):
        now = datetime.now()
        stats = {}
        
        # Filtere die Daten nach Zeiträumen
        day_data = [entry['players'] for entry in self.stats_db['hourly'] 
                    if (now - entry['timestamp']).total_seconds() <= 86400]  # 24 Stunden
        week_data = [entry['players'] for entry in self.stats_db['hourly']
                     if (now - entry['timestamp']).total_seconds() <= 604800]  # 7 Tage
        month_data = [entry['players'] for entry in self.stats_db['hourly']
                      if (now - entry['timestamp']).total_seconds() <= 2592000]  # 30 Tage
        
        # Berechne Statistiken für jeden Zeitraum
        def calc_stats(data):
            if not data:
                return {'avg': 0, 'max': 0, 'min': 0}
            return {
                'avg': round(sum(data) / len(data), 1),
                'max': max(data),
                'min': min(data)
            }
        
        day_stats = calc_stats(day_data)
        week_stats = calc_stats(week_data)
        month_stats = calc_stats(month_data)
        
        return {
            'day_avg': day_stats['avg'],
            'day_max': day_stats['max'],
            'day_min': day_stats['min'],
            'week_avg': week_stats['avg'],
            'week_max': week_stats['max'],
            'week_min': week_stats['min'],
            'month_avg': month_stats['avg'],
            'month_max': month_stats['max'],
            'month_min': month_stats['min']
        }

    async def send_webhook(self, embed):
        webhook = discord.Webhook.from_url(
            self.webhook_url,
            session=self.session
        )
        await webhook.send(embed=embed)

    def __del__(self):
        if hasattr(self, 'hourly_update_task') and self.hourly_update_task:
            self.hourly_update_task.cancel()
        if hasattr(self, 'driver') and self.driver:
            self.driver.quit()

File: bot.js
const Discord = require("discord.js");
const puppeteer = require("puppeteer");
const fs = require("fs").promises;
const { UPDATE_INTERVAL, WEBHOOK_URL, SERVER_ID } = require("./config");

class ServerStatsBot {
  constructor(webhookUrl, serverId) {
    this.webhookUrl = webhookUrl;
    this.serverId = serverId;
    this.statsFile = "stats_db.json";
    this.statsDb = { hourly: [] };
    this.browser = null;
    this.updateInterval = null;
  }

  async init() {
    try {
      await this.loadStats();
      await this.setupBrowser();
      console.log("Bot initialisiert");
    } catch (error) {
      console.error("Fehler bei der Initialisierung:", error);
    }
  }

  async setupBrowser() {
    try {
      this.browser = await puppeteer.launch({
        headless: "new",
        args: ["--no-sandbox", "--disable-dev-shm-usage", "--disable-gpu"],
      });
      console.log("Browser erfolgreich initialisiert");
    } catch (error) {
      console.error("Fehler beim Browser-Setup:", error);
      this.browser = null;
    }
  }

  async getPlayerCount() {
    try {
      if (!this.browser) {
        await this.setupBrowser();
      }

      const page = await this.browser.newPage();
      await page.setUserAgent(
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36"
      );

      await page.goto("https://servers.redm.gg/servers/detail/bzy79l", {
        waitUntil: "networkidle0",
      });

      const playerCount = await page.evaluate(() => {
        const element = document.querySelector("div.connect-bar div.right");
        if (element) {
          const match = element.textContent.match(/\d+/);
          return match ? parseInt(match[0]) : 0;
        }
        return 0;
      });

      await page.close();
      return playerCount;
    } catch (error) {
      console.error("Fehler beim Scraping:", error);
      return 0;
    }
  }

  async loadStats() {
    try {
      const data = await fs.readFile(this.statsFile, "utf8");
      const stats = JSON.parse(data);
      stats.hourly = stats.hourly.map((entry) => ({
        ...entry,
        timestamp: new Date(entry.timestamp),
      }));
      this.statsDb = stats;
    } catch (error) {
      this.statsDb = { hourly: [] };
    }
  }

  async saveStats() {
    try {
      await fs.writeFile(this.statsFile, JSON.stringify(this.statsDb));
    } catch (error) {
      console.error("Fehler beim Speichern der Statistiken:", error);
    }
  }

  calculateStats() {
    const now = new Date();

    const filterData = (hours) => {
      return this.statsDb.hourly
        .filter((entry) => now - new Date(entry.timestamp) <= hours * 3600000)
        .map((entry) => entry.players);
    };

    const calcStats = (data) => {
      if (!data.length) return { avg: 0, max: 0, min: 0 };
      return {
        avg: Math.round((data.reduce((a, b) => a + b, 0) / data.length) * 10) / 10,
        max: Math.max(...data),
        min: Math.min(...data),
      };
    };

    const dayData = filterData(24);
    const weekData = filterData(168);
    const monthData = filterData(720);

    const dayStats = calcStats(dayData);
    const weekStats = calcStats(weekData);
    const monthStats = calcStats(monthData);

    return {
      day_avg: dayStats.avg,
      day_max: dayStats.max,
      day_min: dayStats.min,
      week_avg: weekStats.avg,
      week_max: weekStats.max,
      week_min: weekStats.min,
      month_avg: monthStats.avg,
      month_max: monthStats.max,
      month_min: monthStats.min,
    };
  }

  async sendWebhook(players, stats) {
    const webhook = new Discord.WebhookClient({ url: this.webhookUrl });

    const embed = new Discord.EmbedBuilder()
      .setTitle("Misty Mountain - Aktuelle Spielerzahl")
      .setDescription(`**${players}** Spieler online`)
      .setColor(0x0099ff)
      .setTimestamp()
      .addFields(
        {
          name: "24 Stunden",
          value: `Durchschnitt: ${stats.day_avg}\nMaximum: ${stats.day_max}\nMinimum: ${stats.day_min}`,
          inline: true,
        },
        {
          name: "7 Tage",
          value: `Durchschnitt: ${stats.week_avg}\nMaximum: ${stats.week_max}\nMinimum: ${stats.week_min}`,
          inline: true,
        },
        {
          name: "30 Tage",
          value: `Durchschnitt: ${stats.month_avg}\nMaximum: ${stats.month_max}\nMinimum: ${stats.month_min}`,
          inline: true,
        }
      );

    await webhook.send({ embeds: [embed] });
  }

  async update() {
    try {
      console.log(`\nFühre Update aus - ${new Date().toLocaleTimeString()}`);
      const players = await this.getPlayerCount();

      this.statsDb.hourly.push({
        timestamp: new Date(),
        players,
      });

      const stats = this.calculateStats();
      await this.sendWebhook(players, stats);
      await this.saveStats();

      // Alte Einträge entfernen (älter als 30 Tage)
      const now = new Date();
      this.statsDb.hourly = this.statsDb.hourly.filter((entry) => now - new Date(entry.timestamp) <= 2592000000);

      console.log(`Update abgeschlossen - ${players} Spieler online`);
    } catch (error) {
      console.error("Fehler beim Update:", error);
    }
  }

  async start() {
    await this.init();
    await this.update();
    this.updateInterval = setInterval(() => this.update(), UPDATE_INTERVAL * 1000);
    console.log(`Bot läuft - Update-Intervall: ${UPDATE_INTERVAL} Sekunden`);
  }

  async stop() {
    if (this.updateInterval) {
      clearInterval(this.updateInterval);
    }
    if (this.browser) {
      await this.browser.close();
    }
    console.log("Bot wurde beendet");
  }
}

module.exports = ServerStatsBot;
